File: safe_copy.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');

// 1. Parse and validate arguments
const args = process.argv.slice(2);
if (args.length < 2 || args.length > 3) {
    console.error("Error: Invalid number of arguments.");
    console.error(`Usage: node ${path.basename(process.argv[1])} <source_directory> <destination_directory> [overwrite]`);
    process.exit(1);
}

const srcDir = path.resolve(args[0]);
const destDir = path.resolve(args[1]);
const overwrite = args[2] === 'true';

// 2. Setup timestamped log file
const timestamp = new Date().toISOString().replace(/T/, '_').replace(/\..+/, '').replace(/:/g, '-');
const logFile = path.join(process.cwd(), `copy_log_${timestamp}.log`);

function logMessage(message) {
    const time = new Date().toISOString().replace(/T/, ' ').replace(/\..+/, '');
    const logLine = `[${time}] ${message}\n`;
    process.stdout.write(logLine);
    fs.appendFileSync(logFile, logLine);
}

// 3. Pre-flight validations
if (!fs.existsSync(srcDir) || !fs.statSync(srcDir).isDirectory()) {
    console.error(`Error: Source directory '${srcDir}' does not exist.`);
    process.exit(1);
}

try {
    fs.accessSync(srcDir, fs.constants.R_OK);
} catch (e) {
    console.error(`Error: Missing read permissions on source directory '${srcDir}'.`);
    process.exit(1);
}

if (!fs.existsSync(destDir) || !fs.statSync(destDir).isDirectory()) {
    console.error(`Error: Destination directory '${destDir}' does not exist.`);
    process.exit(1);
}

try {
    fs.accessSync(destDir, fs.constants.W_OK);
} catch (e) {
    console.error(`Error: Missing write permissions on destination directory '${destDir}'.`);
    process.exit(1);
}

const files = fs.readdirSync(srcDir).filter(file => {
    return fs.statSync(path.join(srcDir, file)).isFile();
});

if (files.length === 0) {
    console.error(`Error: Source directory '${srcDir}' is empty.`);
    process.exit(1);
}

// 4. Execute file copying
logMessage(`Starting file copy from '${srcDir}' to '${destDir}'`);
let successCount = 0;
let skipCount = 0;
let failCount = 0;

for (const file of files) {
    const srcFile = path.join(srcDir, file);
    const destFile = path.join(destDir, file);

    if (fs.existsSync(destFile)) {
        if (!overwrite) {
            logMessage(`Skipped: ${file} already exists in destination`);
            skipCount++;
            continue;
        }
    }

    try {
        // copyFileSync preserves fundamental contents; native Node limits direct metadata adjustments 
        fs.copyFileSync(srcFile, destFile, fs.constants.COPYFILE_FICLONE);
        
        // Preserve modification and access timestamps
        const stat = fs.statSync(srcFile);
        fs.utimesSync(destFile, stat.atime, stat.mtime);
        
        logMessage(`Copied: ${file}`);
        successCount++;
    } catch (err) {
        logMessage(`Failed to copy: ${file}`);
        failCount++;
    }
}

logMessage("File copy operation completed.");
logMessage(`Summary -> Successful: ${successCount} | Skipped: ${skipCount} | Failed: ${failCount}`);

process.exit(failCount > 0 ? 1 : 0);

File: safe_copy.py
#!/usr/bin/env python3
import sys
import os
import shutil
from datetime import datetime

# 1. Parse and validate arguments
if len(sys.argv) < 3 or len(sys.argv) > 4:
    print("Error: Invalid number of arguments.", file=sys.stderr)
    print(f"Usage: {sys.argv[0]} <source_directory> <destination_directory> [overwrite]", file=sys.stderr)
    sys.exit(1)

src_dir = os.path.abspath(sys.argv[1])
dest_dir = os.path.abspath(sys.argv[2])
overwrite = sys.argv[3].lower() == 'true' if len(sys.argv) == 4 else False

# 2. Setup timestamped log file
log_filename = f"copy_log_{datetime.now().strftime('%Y-%m-%d_%H:%M:%S')}.log"

def log_message(message):
    timestamp = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
    log_line = f"[{timestamp}] {message}\n"
    sys.stdout.write(log_line)
    with open(log_filename, "a", encoding="utf-8") as log_file:
        log_file.write(log_line)

# 3. Pre-flight validations
if not os.path.isdir(src_dir):
    print(f"Error: Source directory '{src_dir}' does not exist.", file=sys.stderr)
    sys.exit(1)

if not os.access(src_dir, os.R_OK):
    print(f"Error: Missing read permissions on source directory '{src_dir}'.", file=sys.stderr)
    sys.exit(1)

if not os.path.isdir(dest_dir):
    print(f"Error: Destination directory '{dest_dir}' does not exist.", file=sys.stderr)
    sys.exit(1)

if not os.access(dest_dir, os.W_OK):
    print(f"Error: Missing write permissions on destination directory '{dest_dir}'.", file=sys.stderr)
    sys.exit(1)

# Extract only files
files = [f for f in os.listdir(src_dir) if os.path.isfile(os.path.join(src_dir, f))]

if not files:
    print(f"Error: Source directory '{src_dir}' is empty.", file=sys.stderr)
    sys.exit(1)

# 4. File Copy Execution
log_message(f"Starting file copy from '{src_dir}' to '{dest_dir}'")
success_count = 0
skip_count = 0
fail_count = 0

for file in files:
    src_file = os.path.join(src_dir, file)
    dest_file = os.path.join(dest_dir, file)

    if os.path.exists(dest_file):
        if not overwrite:
            log_message(f"Skipped: {file} already exists in destination")
            skip_count += 1
            continue

    try:
        # copy2 preserves permissions, modifications, metadata attributes
        shutil.copy2(src_file, dest_file)
        log_message(f"Copied: {file}")
        success_count += 1
    except Exception:
        log_message(f"Failed to copy: {file}")
        fail_count += 1

log_message("File copy operation completed.")
log_message(f"Summary -> Successful: {success_count} | Skipped: {skip_count} | Failed: {fail_count}")

sys.exit(1 if fail_count > 0 else 0)
